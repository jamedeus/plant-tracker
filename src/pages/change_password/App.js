import React, { useState, useMemo, useRef } from 'react';
import { useTheme } from 'src/context/ThemeContext';
import Navbar from 'src/components/Navbar';
import Cookies from 'js-cookie';
import clsx from 'clsx';

function App() {
    const { ToggleThemeOption } = useTheme();

    const formRef = useRef(null);
    const [oldPassword, setOldPassword] = useState('');
    const [newPassword1, setNewPassword1] = useState('');
    const [newPassword2, setNewPassword2] = useState('');
    const [oldPasswordIncorrect, setOldPasswordIncorrect] = useState(false);
    const [newPasswordError, setNewPasswordError] = useState(false);

    const submit = async (e) => {
        e.preventDefault();
        // Post old and new password to backend
        const response = await fetch("/accounts/change_password/submit/", {
            method: "POST",
            headers: {
                "Content-Type": "application/x-www-form-urlencoded",
                "X-CSRFToken": Cookies.get('csrftoken')
            },
            credentials: "include",
            body: new URLSearchParams({
                old_password: oldPassword,
                new_password1: newPassword1,
                new_password2: newPassword2,
            }),
        });
        // Redirect to overview if logged in successfully
        if (response.ok) {
            window.location.href = '/';
        // Show correct error message if login failed
        } else {
            const errors = await response.json();
            if (errors.errors.old_password) {
                setOldPasswordIncorrect(true);
            }
            if (errors.errors.new_password2) {
                setNewPasswordError(errors.errors.new_password2[0]);
            }
        }
    };

    const DropdownMenuOptions = useMemo(() => (
        <>
            <li><a onClick={() => window.location.href = "/"}>
                Overview
            </a></li>
            <ToggleThemeOption />
        </>
    ), [ToggleThemeOption]);

    return (
        <div className="container flex flex-col h-screen mx-auto items-center">
            <Navbar
                menuOptions={DropdownMenuOptions}
                title="Change Password"
            />

            <form ref={formRef} className="flex flex-col gap-4 mt-[15vh]">
                <label className="form-control w-full">
                    <div className="label">
                        <span className="label-text">Old password</span>
                    </div>
                    <input
                        name="old_password"
                        type="password"
                        className={clsx(
                            "input w-full input-bordered",
                            oldPasswordIncorrect && "input-error"
                        )}
                        value={oldPassword}
                        onInput={() => setOldPasswordIncorrect(false)}
                        onChange={(e) => setOldPassword(e.target.value)}
                    />
                </label>
                {oldPasswordIncorrect && (
                    <span className="text-error">
                        Old password incorrect
                    </span>
                )}
                <label className="form-control w-full relative">
                    <div className="label">
                        <span className="label-text">New password</span>
                    </div>
                    <input
                        name="new_password1"
                        type="password"
                        className={clsx(
                            "input w-full input-bordered",
                            newPasswordError && "input-error"
                        )}
                        value={newPassword1}
                        onChange={(e) => setNewPassword1(e.target.value)}
                    />
                </label>
                <label className="form-control w-full">
                    <div className="label">
                        <span className="label-text">Confirm new password</span>
                    </div>
                    <input
                        name="new_password2"
                        type="password"
                        className={clsx(
                            "input w-full input-bordered",
                            newPasswordError && "input-error"
                        )}
                        value={newPassword2}
                        onInput={() => setNewPasswordError(false)}
                        onChange={(e) => setNewPassword2(e.target.value)}
                    />
                </label>
                {newPasswordError && (
                    <span className="text-error">
                        {newPasswordError}
                    </span>
                )}

                <button
                    className="btn btn-success mt-6"
                    onClick={(e) => submit(e)}
                >
                    Change password
                </button>
            </form>
        </div>
    );
}

export default App;
