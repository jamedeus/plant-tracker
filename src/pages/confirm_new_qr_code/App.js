import React, { useEffect, useMemo } from 'react';
import PropTypes from 'prop-types';
import { sendPostRequest, parseDomContext } from 'src/util';
import Navbar from 'src/components/Navbar';
import NavbarDropdownOptions from 'src/components/NavbarDropdownOptions';
import PlantDetails from 'src/components/PlantDetails.js';
import GroupDetails from 'src/components/GroupDetails.js';
import { openErrorModal } from 'src/components/ErrorModal';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faXmark, faCheck } from '@fortawesome/free-solid-svg-icons';

const Layout = ({ type, thumbnailUrl, handleConfirm }) => {
    return (
        <div className="flex flex-col gap-8 text-center my-auto">
            <p className="text-lg font-bold">
                Is this the new QR code for your {type}?
            </p>
            {thumbnailUrl && (
                <div className="mx-auto p-4 bg-base-200 rounded-3xl">
                    <img
                        className="max-h-[50vh] rounded-xl object-contain"
                        src={thumbnailUrl}
                    />
                </div>
            )}

            {/* Confirm/cancel buttons */}
            <div className="flex gap-4 mx-auto mb-8">
                <button
                    className="btn btn-error btn-square text-white"
                    onClick={() => window.location.href = "/"}
                    title="Don't change QR code"
                >
                    <FontAwesomeIcon className="size-6" icon={faXmark} />
                </button>
                <button
                    className="btn btn-success btn-square text-white"
                    onClick={handleConfirm}
                    title="Change QR code"
                >
                    <FontAwesomeIcon className="size-6" icon={faCheck} />
                </button>
            </div>
        </div>
    );
};

Layout.propTypes = {
    type: PropTypes.oneOf([
        'plant',
        'group'
    ]).isRequired,
    thumbnailUrl: PropTypes.string,
    handleConfirm: PropTypes.func.isRequired
};

function App() {
    // Load UUIDs and plant/group state from django template context
    const type = parseDomContext("type");
    const instance = parseDomContext("instance");
    const newUuid = parseDomContext("new_uuid");

    // Reload if user navigates to page by pressing back button (change QR code
    // request likely expired, replace with register or manage page)
    useEffect(() => {
        const handleBackButton = async (event) => {
            if (event.persisted) {
                window.location.reload();
            }
        };

        // Add listener on mount, remove on unmount
        window.addEventListener('pageshow', handleBackButton);
        return () => {
            window.removeEventListener('pageshow', handleBackButton);
        };
    }, []);

    const handleConfirm = async () => {
        const response = await sendPostRequest(
            '/change_uuid',
            {
                uuid: instance.uuid,
                new_id: newUuid
            }
        );
        // Reload page if changed successfully
        if (response.ok) {
            window.location.reload();
        } else {
            const error = await response.json();
            openErrorModal(JSON.stringify(error));
        }
    };

    // Plant/group details card shown when title is clicked
    const DetailsDropdown = useMemo(() => (
        <div className="details-card">
            <div className="card-body">
                {type === 'plant' ? (
                    <PlantDetails
                        species={instance.species}
                        pot_size={instance.pot_size}
                        description={instance.description}
                    />
                ) : (
                    <GroupDetails
                        location={instance.location}
                        description={instance.description}
                    />
                )}
            </div>
        </div>
    ), []);

    return (
        <div className="container flex flex-col mx-auto h-screen">
            <Navbar
                menuOptions={<NavbarDropdownOptions />}
                title={instance.display_name}
                titleOptions={DetailsDropdown}
            />
            <Layout
                type={type}
                thumbnailUrl={instance.thumbnail}
                handleConfirm={handleConfirm}
            />
        </div>
    );
}

export default App;
