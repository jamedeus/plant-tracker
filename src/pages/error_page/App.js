import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import Navbar from 'src/components/Navbar';
import NavbarDropdownOptions from 'src/components/NavbarDropdownOptions';

function App({ errorMessage }) {
    return (
        <div
            className="container flex flex-col full-screen mx-auto items-center gap-16"
            data-testid="error-page"
        >
            <Navbar
                menuOptions={<NavbarDropdownOptions />}
                title="Error"
                showScanButton={false}
            />

            <div className="flex flex-col text-center text-lg mt-[15vh]">
                {errorMessage}
            </div>

            <Link to="/" className='btn btn-accent' discover="none">
                Go to Overview
            </Link>
        </div>
    );
}

export default App;

App.propTypes = {
    errorMessage: PropTypes.string.isRequired
};
