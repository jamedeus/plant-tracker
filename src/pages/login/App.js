import React, { useState, useMemo, useRef } from 'react';
import { useTheme } from 'src/context/ThemeContext';
import Navbar from 'src/components/Navbar';
import Cookies from 'js-cookie';
import clsx from 'clsx';

function App() {
    const { ToggleThemeOption } = useTheme();

    const formRef = useRef(null);
    const [username, setUsername] = useState('');
    const [password, setPassword] = useState('');
    const [showError, setShowError] = useState(false);

    const submit = async (e) => {
        e.preventDefault();
        // Post FormData to backend
        const formData = new FormData(formRef.current);
        const response = await fetch('/accounts/login/', {
            method: 'POST',
            body: formData,
            headers: {
                'Accept': 'application/json, text/plain, */*',
                "X-CSRFToken": Cookies.get('csrftoken')
            }
        });
        // Redirect if logged in successfully
        if (response.ok) {
            // Redirect to url in querystring if present
            const params = new URL(window.location.href).searchParams;
            if (params.get('next')) {
                window.location.href = params.get('next');
            // Redirect to overview if no querystring
            } else {
                window.location.href = '/';
            }
        // Show error text if login failed
        } else {
            setShowError(true);
        }
    };

    const DropdownMenuOptions = useMemo(() => (
        <>
            <li><a onClick={() => window.location.href = "/"}>
                Overview
            </a></li>
            <ToggleThemeOption />
        </>
    ), [ToggleThemeOption]);

    return (
        <div className="container flex flex-col h-screen mx-auto items-center">
            <Navbar
                menuOptions={DropdownMenuOptions}
                title="Login"
            />

            <form ref={formRef} className="flex flex-col gap-4 mt-[15vh]">
                <label className="form-control w-full">
                    <div className="label">
                        <span className="label-text">Username</span>
                    </div>
                    <input
                        name="username"
                        type="text"
                        autoCapitalize="off"
                        className={clsx(
                            "input w-full input-bordered",
                            showError && "input-error"
                        )}
                        value={username}
                        onInput={() => setShowError(false)}
                        onChange={e => setUsername(e.target.value)}
                    />
                </label>
                <label className="form-control w-full relative">
                    <div className="label">
                        <span className="label-text">Password</span>
                    </div>
                    <input
                        name="password"
                        type="password"
                        className={clsx(
                            "input w-full input-bordered",
                            showError && "input-error"
                        )}
                        value={password}
                        onInput={() => setShowError(false)}
                        onChange={e => setPassword(e.target.value)}
                    />
                </label>

                <button
                    className="btn btn-success mt-6"
                    onClick={(e) => submit(e)}
                    disabled={username.length < 1 || password.length < 1}
                >
                    Login
                </button>

                {showError && (
                    <span className="text-error text-center">
                        Invalid username or password
                    </span>
                )}
            </form>
        </div>
    );
}

export default App;
