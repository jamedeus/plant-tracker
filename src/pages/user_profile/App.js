import React, { useState, useMemo, useRef, memo } from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import ToggleThemeOption from 'src/components/ToggleThemeOption';
import { EMAIL_REGEX } from 'src/regex';
import { sendPostRequest } from 'src/util';
import { timestampToRelative, timestampToReadable } from 'src/timestampUtils';
import Navbar from 'src/components/Navbar';
import { showToast } from 'src/components/Toast';
import QrScannerButton from 'src/components/QrScannerButton';
import Cookies from 'js-cookie';
import clsx from 'clsx';
import { FaCheck } from 'react-icons/fa6';

const UserDetails = memo(function UserDetails({ initialUserDetails }) {
    // Initialize from SPA-provided state
    const [userDetails, setUserDetails] = useState(initialUserDetails);

    const [firstName, setFirstName] = useState(userDetails.first_name);
    const [lastName, setLastName] = useState(userDetails.last_name);
    const [email, setEmail] = useState(userDetails.email);

    // Enable submit button when 1 or more field has changed
    const submitButtonDisabled = firstName === userDetails.first_name &&
                                 lastName === userDetails.last_name &&
                                 email === userDetails.email ||
                                 !EMAIL_REGEX.test(email);

    const submit = async () => {
        const payload = {
            first_name: firstName,
            last_name: lastName,
            email: email
        };
        const response = await sendPostRequest(
            '/accounts/edit_user_details/',
            payload
        );
        if (response.ok) {
            const data = await response.json();
            setUserDetails(data.user_details);
            showToast('Details updated!', 'green', 2000);
        } else {
            showToast('Unable to update details', 'red', 2000);
        }
    };

    const submitOnEnterKey = async (event) => {
        if (event.key === 'Enter' && !submitButtonDisabled) {
            await submit();
        }
    };

    const resendVerificationEmail = async () => {
        const response = await fetch('/accounts/resend_verification_email/');
        if (response.ok) {
            showToast('Verification email sent!', 'green', 2000);
        } else {
            showToast('Unable to send verification email', 'red', 2000);
        }
    };

    return (
        <div className="flex flex-col gap-4">
            <div className="grid grid-cols-min-max gap-4">
                <div className="my-auto text-nowrap">
                    First Name:
                </div>
                <input
                    name="first_name"
                    type="text"
                    aria-label="First name input"
                    className="input w-full"
                    value={firstName}
                    onChange={(e) => setFirstName(e.target.value)}
                    onKeyDown={(e) => submitOnEnterKey(e)}
                    data-testid="first_name_input"
                />
                <div className="my-auto text-nowrap">
                    Last Name:
                </div>
                <input
                    name="last_name"
                    type="text"
                    aria-label="Last name input"
                    className="input w-full"
                    value={lastName}
                    onChange={(e) => setLastName(e.target.value)}
                    onKeyDown={(e) => submitOnEnterKey(e)}
                    data-testid="last_name_input"
                />
                <div className="my-auto text-nowrap">
                    Email:
                </div>
                <div>
                    <input
                        name="email"
                        type="email"
                        autoCapitalize="off"
                        aria-label="Email address input"
                        className="input w-full"
                        value={email}
                        onChange={(e) => setEmail(e.target.value)}
                        onKeyDown={(e) => submitOnEnterKey(e)}
                        data-testid="email_input"
                    />
                    {userDetails.email_verified ? (
                        <div className="text-success text-sm text-center mt-2">
                            <FaCheck className="fa-inline mr-2" />
                            <span>Email verified</span>
                        </div>
                    ) : (
                        <div className="text-error text-sm text-center mt-2">
                            <span>Not verified â€” </span>
                            <span
                                className="underline cursor-pointer"
                                onClick={resendVerificationEmail}
                            >
                                resend email
                            </span>
                        </div>
                    )}
                </div>
                <div className="my-auto text-nowrap">
                    Joined:
                </div>
                <div className="text-center">
                    <span className="text-nowrap">
                        {timestampToReadable(userDetails.date_joined).split('-')[1]}
                    </span>
                    {/* Space allows line wrapping on very small screens */}
                    <span> </span>
                    <span className="text-nowrap">
                        ({timestampToRelative(userDetails.date_joined)})
                    </span>
                </div>
            </div>
            <button
                className="btn btn-accent mt-4"
                disabled={submitButtonDisabled}
                onClick={submit}
            >
                Save Changes
            </button>
        </div>
    );
});

UserDetails.propTypes = {
    initialUserDetails: PropTypes.shape({
        username: PropTypes.string.isRequired,
        email: PropTypes.string.isRequired,
        email_verified: PropTypes.bool.isRequired,
        first_name: PropTypes.string.isRequired,
        last_name: PropTypes.string.isRequired,
        date_joined: PropTypes.string.isRequired,
    }).isRequired
};

const ChangePassword = memo(function ChangePassword() {
    const formRef = useRef(null);
    const [oldPassword, setOldPassword] = useState('');
    const [newPassword1, setNewPassword1] = useState('');
    const [newPassword2, setNewPassword2] = useState('');
    const [oldPasswordIncorrect, setOldPasswordIncorrect] = useState(false);
    const [newPasswordError, setNewPasswordError] = useState(false);

    // Enable submit button when all fields reach minimum password length and
    // both new password fields match
    const submitButtonDisabled = oldPassword.length < 8 ||
                                 newPassword1.length < 8 ||
                                 newPassword2.length < 8 ||
                                 newPassword1 != newPassword2;

    const submit = async (e) => {
        e.preventDefault();
        // Post old and new password to backend
        const response = await fetch("/accounts/change_password/", {
            method: "POST",
            headers: {
                "Content-Type": "application/x-www-form-urlencoded",
                "X-CSRFToken": Cookies.get('csrftoken')
            },
            credentials: "include",
            body: new URLSearchParams({
                old_password: oldPassword,
                new_password1: newPassword1,
                new_password2: newPassword2,
            }),
        });
        if (response.ok) {
            showToast('Password changed!', 'green', 2000);
        } else {
            const errors = await response.json();
            if (errors.errors.old_password) {
                setOldPasswordIncorrect(true);
            }
            if (errors.errors.new_password2) {
                setNewPasswordError(errors.errors.new_password2[0]);
            }
        }
    };

    const submitOnEnterKey = async (event) => {
        if (event.key === 'Enter' && !submitButtonDisabled) {
            await submit(event);
        }
    };

    return (
        <form ref={formRef} className="flex flex-col gap-2">
            <label title="Must be at least 8 characters, can't be all numbers">
                <span>Old password</span>
                <input
                    name="old_password"
                    type="password"
                    className={clsx(
                        "input w-full",
                        oldPasswordIncorrect && "border-error"
                    )}
                    value={oldPassword}
                    onKeyDown={(e) => submitOnEnterKey(e)}
                    onInput={() => setOldPasswordIncorrect(false)}
                    onChange={(e) => setOldPassword(e.target.value)}
                />
            </label>
            {oldPasswordIncorrect && (
                <span className="text-error text-center">
                    Old password incorrect
                </span>
            )}
            <label title="Must be at least 8 characters, can't be all numbers">
                <span>New password</span>
                <input
                    name="new_password1"
                    type="password"
                    className={clsx(
                        "input w-full",
                        newPasswordError && "border-error"
                    )}
                    value={newPassword1}
                    onKeyDown={(e) => submitOnEnterKey(e)}
                    onChange={(e) => setNewPassword1(e.target.value)}
                />
            </label>
            <label title="Must be at least 8 characters, can't be all numbers">
                <span>Confirm new password</span>
                <input
                    name="new_password2"
                    type="password"
                    className={clsx(
                        "input w-full",
                        newPasswordError && "border-error"
                    )}
                    value={newPassword2}
                    onKeyDown={(e) => submitOnEnterKey(e)}
                    onChange={(e) => setNewPassword2(e.target.value)}
                />
            </label>
            {newPasswordError && (
                <span className="text-error text-center">
                    {newPasswordError}
                </span>
            )}

            <button
                className="btn btn-accent mt-6"
                disabled={submitButtonDisabled}
                onClick={(e) => submit(e)}
            >
                Change Password
            </button>
        </form>
    );
});

const Section = ({ title, open=false, children }) => {
    return (
        <div
            className="collapse collapse-arrow bg-base-200 rounded-2xl"
            tabIndex={0}
        >
            <input
                type="radio"
                name="my-accordion-2"
                defaultChecked={open}
                aria-label={`Show ${title} form`}
            />
            <div className="collapse-title text-xl font-medium text-center p-4">
                {title}
            </div>
            <div className="collapse-content">
                {children}
            </div>
        </div>
    );
};

Section.propTypes = {
    title: PropTypes.string.isRequired,
    open: PropTypes.bool,
    children: PropTypes.node.isRequired
};

function App({ initialUserDetails }) {
    const DropdownMenuOptions = useMemo(() => (
        <>
            <li><Link to='/'>
                Overview
            </Link></li>
            <ToggleThemeOption />
        </>
    ), [ToggleThemeOption]);

    return (
        <div className="container flex flex-col full-screen mx-auto items-center">
            <Navbar
                menuOptions={DropdownMenuOptions}
                title="User Profile"
                topRightButton={<QrScannerButton />}
            />
            <div className="flex flex-col w-96 max-w-[100vw] gap-4 md:gap-6 px-4 md:mt-16">
                <Section title="Details" open={true}>
                    <UserDetails initialUserDetails={initialUserDetails} />
                </Section>
                <Section title="Change Password">
                    <ChangePassword />
                </Section>
                <Link
                    className="btn btn-soft btn-error mx-4 mb-8"
                    to="/accounts/logout/"
                >
                    Log Out
                </Link>
            </div>
        </div>
    );
}

export default App;

App.propTypes = {
    initialUserDetails: PropTypes.shape({
        username: PropTypes.string.isRequired,
        email: PropTypes.string.isRequired,
        email_verified: PropTypes.bool.isRequired,
        first_name: PropTypes.string.isRequired,
        last_name: PropTypes.string.isRequired,
        date_joined: PropTypes.string.isRequired,
    }).isRequired
};
