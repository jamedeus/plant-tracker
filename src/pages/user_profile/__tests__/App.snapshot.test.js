import createMockContext from 'src/testUtils/createMockContext';
import App from '../App';

describe('App', () => {
    it('matches snapshot', () => {
        // Render App, confirm matches snapshot
        createMockContext('user_accounts_enabled', true);
        const { container } = render(<App initialState={{
            user_details: {
                username: "cdanger",
                email: "totally.not.anthony.weiner@gmail.com",
                first_name: "Carlos",
                last_name: "Danger",
                date_joined: "2025-04-06T00:08:53.392806+00:00"
            }
        }} />);
        expect(container).toMatchSnapshot();
    });
});
