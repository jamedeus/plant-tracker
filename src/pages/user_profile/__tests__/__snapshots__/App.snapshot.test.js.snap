// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App matches snapshot 1`] = `
<div>
  <div
    class="container flex flex-col full-screen mx-auto items-center"
    data-testid="user-profile-page"
  >
    <div
      class="navbar bg-base-100 fixed top-0 z-99"
    >
      <div
        class="dropdown justify-start min-w-12"
      >
        <div
          aria-label="Navigation menu"
          class="btn btn-ghost btn-circle size-12"
          role="button"
          tabindex="0"
        >
          <svg
            class="size-5"
            fill="none"
            stroke="currentColor"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M4 6h16M4 12h16M4 18h7"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
          </svg>
        </div>
        <ul
          class="dropdown-content menu z-90 shadow-sm bg-base-300 rounded-box mt-3"
          tabindex="0"
        >
          <li>
            <a
              href="/"
            >
              Overview
            </a>
          </li>
          <li>
            <button>
              Light mode
            </button>
          </li>
        </ul>
      </div>
      <div
        class="mx-auto shrink min-w-0"
        title="Scroll to top"
      >
        <div
          class="dropdown dropdown-center w-full"
        >
          <a
            aria-level="1"
            class="w-full px-0 text-nowrap font-semibold"
            role="heading"
            tabindex="-1"
          >
            <span
              class="px-4 leading-[2.75rem] truncate"
              style="font-size: 14px;"
            >
              User Profile
            </span>
          </a>
        </div>
      </div>
      <div
        class="justify-end min-w-12"
      >
        <button
          aria-label="Open QR scanner"
          class="btn btn-ghost btn-circle size-12"
          title="Open QR scanner"
        >
          <svg
            class="size-6"
            fill="none"
            height="1em"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M3 7V5a2 2 0 0 1 2-2h2"
            />
            <path
              d="M17 3h2a2 2 0 0 1 2 2v2"
            />
            <path
              d="M21 17v2a2 2 0 0 1-2 2h-2"
            />
            <path
              d="M7 21H5a2 2 0 0 1-2-2v-2"
            />
            <circle
              cx="12"
              cy="12"
              r="3"
            />
            <path
              d="m16 16-1.9-1.9"
            />
          </svg>
        </button>
      </div>
    </div>
    <div
      class="flex flex-col w-96 max-w-[100vw] gap-4 md:gap-6 px-4 md:mt-16"
    >
      <div
        class="collapse collapse-arrow bg-base-200 rounded-2xl"
        tabindex="0"
      >
        <input
          aria-label="Show Details form"
          checked=""
          name="my-accordion-2"
          type="radio"
        />
        <div
          class="collapse-title text-xl font-medium text-center p-4"
        >
          Details
        </div>
        <div
          class="collapse-content"
        >
          <div
            class="flex flex-col gap-4"
          >
            <div
              class="grid grid-cols-min-max gap-4"
            >
              <div
                class="my-auto text-nowrap"
              >
                First Name:
              </div>
              <input
                aria-label="First name input"
                class="input w-full"
                data-testid="first_name_input"
                name="first_name"
                type="text"
                value="Carlos"
              />
              <div
                class="my-auto text-nowrap"
              >
                Last Name:
              </div>
              <input
                aria-label="Last name input"
                class="input w-full"
                data-testid="last_name_input"
                name="last_name"
                type="text"
                value="Danger"
              />
              <div
                class="my-auto text-nowrap"
              >
                Email:
              </div>
              <div>
                <input
                  aria-label="Email address input"
                  autocapitalize="off"
                  class="input w-full"
                  data-testid="email_input"
                  name="email"
                  type="email"
                  value="totally.not.anthony.weiner@gmail.com"
                />
                <div
                  class="text-error text-sm text-center mt-2"
                >
                  <span>
                    Not verified â€” 
                  </span>
                  <span
                    class="underline cursor-pointer"
                  >
                    resend email
                  </span>
                </div>
              </div>
              <div
                class="my-auto text-nowrap"
              >
                Joined:
              </div>
              <div
                class="text-center"
              >
                <span
                  class="text-nowrap"
                >
                   April 5, 2025
                </span>
                <span>
                   
                </span>
                <span
                  class="text-nowrap"
                >
                  (
                  in 1 year
                  )
                </span>
              </div>
            </div>
            <button
              class="btn btn-accent mt-4"
              disabled=""
            >
              Save Changes
            </button>
          </div>
        </div>
      </div>
      <div
        class="collapse collapse-arrow bg-base-200 rounded-2xl"
        tabindex="0"
      >
        <input
          aria-label="Show Change Password form"
          name="my-accordion-2"
          type="radio"
        />
        <div
          class="collapse-title text-xl font-medium text-center p-4"
        >
          Change Password
        </div>
        <div
          class="collapse-content"
        >
          <form
            class="flex flex-col gap-2"
          >
            <label
              title="Must be at least 8 characters, can't be all numbers"
            >
              <span>
                Old password
              </span>
              <input
                class="input w-full"
                name="old_password"
                type="password"
                value=""
              />
            </label>
            <label
              title="Must be at least 8 characters, can't be all numbers"
            >
              <span>
                New password
              </span>
              <input
                class="input w-full"
                name="new_password1"
                type="password"
                value=""
              />
            </label>
            <label
              title="Must be at least 8 characters, can't be all numbers"
            >
              <span>
                Confirm new password
              </span>
              <input
                class="input w-full"
                name="new_password2"
                type="password"
                value=""
              />
            </label>
            <button
              class="btn btn-accent mt-6"
              disabled=""
            >
              Change Password
            </button>
          </form>
        </div>
      </div>
      <a
        class="btn btn-soft btn-error mx-4 mb-8"
        href="/accounts/logout/"
      >
        Log Out
      </a>
    </div>
  </div>
</div>
`;
